### ISSUE TITLE
docs: how to match URL query parameters?

### ISSUE THREAD

**Issue by lucab** [lipanski/mockito#71]
I have an usecase where I want to match the presence of some URL query parameters in the client request. I haven't find any such example in docs and test, and I'm having an hard time writing the Matcher for that.

The scenario is that an endpoint requires multiple URL query parameters (say `a` and `b`), so I want to match `endpoint?a=foo&c=bar&b=baz` and similar permutations (possibly ignoring the actual values).

@lipanski what would be your suggestion to approach this?

**Comment by lipanski** [comment]
@lucab one option would be to use the Regex matcher for the path argument - something like [this](https://docs.rs/mockito/0.17.1/mockito/index.html#example-6) - but applied to query parameters:

```rust
use mockito::{mock, Matcher};

let _m = mock("GET", Matcher::Regex(r"^/endpoint\?a\=(\w+)\&b\=(\w+)".to_string())).create();
```

would this work for you? otherwise I could very well imagine introducing a proper query parameter matching method - like `mock("GET", "/").match_query(...)`, which takes a `HashMap<String, Into<Matcher>>` as an argument and ignores the parameter order.

**Comment by lucab** [comment]
That was my original approach, but after writing and running the regex-matcher (and failing, duh), I realized that permutations are valid too: {a,b,c}, {c,a,b}, {b,a}.
I guess it is theoretically possible to encode this in a regexp (I didn't actually try, though), but it may not be very ergonomic.

**Comment by gakonst** [comment]
> @lucab one option would be to use the Regex matcher for the path argument - something like [this](https://docs.rs/mockito/0.17.1/mockito/index.html#example-6) - but applied to query parameters:
> 
> ```rust
> use mockito::{mock, Matcher};
> 
> let _m = mock("GET", Matcher::Regex(r"^/endpoint\?a\=(\w+)\&b\=(\w+)".to_string())).create();
> ```
> 
> would this work for you? otherwise I could very well imagine introducing a proper query parameter matching method - like `mock("GET", "/").match_query(...)`, which takes a `HashMap<String, Into<Matcher>>` as an argument and ignores the parameter order.

That's my current approach, is there some way to retrieve the URL query arguments and return them in the body, similar to an echo mock?

**Comment by lucab** [comment]
@gakonst I'm curious, how do you handle valid arguments permutations?

**Comment by lipanski** [comment]
@lucab @gakonst I managed to improve the current behaviour by adding `Mock#match_query`, `Matcher::UrlEncoded` and `Matcher::AllOf`. When used in combination, they should simplify the problem of matching query parameters significantly. Please give it a try.

More details here: https://docs.rs/mockito/0.18.0/mockito/#matching-by-query

Released as part of [0.18.0](https://github.com/lipanski/mockito/releases/tag/0.18.0).
